airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/cdc/PostgresCdcCtidInitializer.java

```diff
@@ -211,10 +201,125 @@ public static List<AutoCloseableIterator<AirbyteMessage>> cdcCtidIteratorsCombin
         LOGGER.warn("Streams are under vacuuming, not going to process WAL");
         return initialSyncCtidIterators;
       }
-
     } catch (final SQLException e) {
       throw new RuntimeException(e);
     }
   }
 
+  static class CdcIterator extends AbstractIterator<AirbyteMessage> implements AutoCloseableIterator<AirbyteMessage> {
+
+    private final Iterator<DebeziumComponent.Output> outputIterator;
+    private final StateManager stateManager;
+    private final Instant emittedAt;
+
+    private Iterator<AirbyteMessage> messageIterator;
+
+    public CdcIterator(Iterator<DebeziumComponent.Output> outputIterator, StateManager stateManager, Instant emittedAt) {
+      this.outputIterator = outputIterator;
+      this.stateManager = stateManager;
+      this.emittedAt = emittedAt;
+      this.messageIterator = nextIterator(outputIterator.next());
+    }
+
+    @Override
+    public void close() {
+      endOfData();
+    }
+
+    @Nullable
+    @Override
+    protected AirbyteMessage computeNext() {
+      if (!messageIterator.hasNext()) {
+        if (!outputIterator.hasNext()) {
+          return endOfData();
+        }
+        messageIterator = nextIterator(outputIterator.next());
+      }
+      return messageIterator.next();
+    }
+
+    private Iterator<AirbyteMessage> nextIterator(final DebeziumComponent.Output output) {
+      return new Iterator<>() {
+
+        private Iterator<DebeziumComponent.Record> recordIterator = output.data().iterator();
+
+        @Override
+        public boolean hasNext() {
+          return recordIterator != null;
+        }
+
+        @Override
+        public AirbyteMessage next() {
+          if (recordIterator == null) {
+            throw new NoSuchElementException();
+          }
+          if (recordIterator.hasNext()) {
+            return toAirbyteRecordMessage(recordIterator.next(), emittedAt);
+          } else {
```

Comment:
is this intends to send out the last message as a state message? I think there are more logic to this thus I created `SourceStateIterator` to wrap the record iterator into a state/record iterator, with the counting mechanism we need to deliver as the original goal. Essentially it needs:

1. error handling logic (throw when record iterator throws an exception)
2. checkpointing logic (how often to send an intermediate state message)

I feel CdcIterator could be a good idea, converting DebeziumComponent.Record into State/Record messages, but probably we need to wrap it with SourceStateIterator.

Reply:
Yes and no. The point of the component is that `collect` always produces a "checkpointable batch", i.e. a bunch of `RECORD`s followed by a `STATE` message which respects a checkpointing policy (batch not too big, time elapsed not too long, etc.). In a sense, it's the component's job to implement this checkpointing logic so that its clients don't need to worry about it.